theming.js

```diff
@@ -291,3 +313,303 @@ var ThemeManager = new Lang.Class({
+        this._dockActor = dock.actor;
+        this._dock = dock;
+        this._position = Utils.getPosition(this._settings);
+
+        this._backgroundColor = '0,0,0';
+        this._transparentAlpha = '0.2';
+        this._opaqueAlpha = '1';
+        this._transparentAlphaBorder = '0.1';
+        this._opaqueAlphaBorder = '0.5';
+        this._transparentTransition = '0ms';
+        this._opaqueTransition = '0ms';
+
+        this._updateStyles();
+
+        this._signalsHandler = new Utils.GlobalSignalsHandler();
+        this._injectionsHandler = new Utils.InjectionsHandler();
+        this._trackedWindows = new Map();
+    },
+
+    enable: function() {
+        // ensure I never double-register/inject
+        // although it should never happen
+        this.disable();
+
+        this._signalsHandler.addWithLabel('transparency', [
+            global.window_group,
+            'actor-added',
+            Lang.bind(this, this._onWindowActorAdded)
+        ], [
+            global.window_group,
+            'actor-removed',
+            Lang.bind(this, this._onWindowActorRemoved)
+        ], [
+            global.window_manager,
+            'switch-workspace',
+            Lang.bind(this, this._updateSolidStyle)
+        ],[
+            Main.overview,
+            'hiding',
+            Lang.bind(this, this._updateSolidStyle)
+        ]);
+
+        // Window signals
+        global.get_window_actors().forEach(function(win) {
+            if (win.get_meta_window().get_wm_class() !== 'Gnome-shell')
+                this._addWindowSignals(win);
+        }, this);
+
+        if (this._settings.get_enum('transparency-mode') === TransparencyMode.adaptive)
+            this._enableAdaptive();
+
+        if (this._actor.get_stage())
+            this._updateSolidStyle();
+    },
+
+    disable: function() {
+        this._disableAdaptive();
+
+        // ensure I never double-register/inject
+        // although it should never happen
+        this._signalsHandler.removeWithLabel('transparency');
+
+        for (let key of this._trackedWindows.keys())
+            this._trackedWindows.get(key).forEach(id => {
+                key.disconnect(id);
+            });
+        this._trackedWindows.clear();
+    },
+
+    destroy: function() {
+        this.disable();
+        this._signalsHandler.destroy();
+        this._injectionsHandler.destroy();
+    },
+
+    _addWindowSignals: function(metaWindowActor) {
+        let signalIds = [];
+        ['allocation-changed', 'notify::visible'].forEach(s => {
+            signalIds.push(metaWindowActor.connect(s, Lang.bind(this, this._updateSolidStyle)));
+        });
+        this._trackedWindows.set(metaWindowActor, signalIds);
+    },
+
+    _onWindowActorAdded: function(container, metaWindowActor) {
+        this._addWindowSignals(metaWindowActor);
+    },
+
+    _onWindowActorRemoved: function(container, metaWindowActor) {
+        this._trackedWindows.get(metaWindowActor).forEach(id => {
```

Comment:
We need to check that this._trackedWindows.get() is not undefined.  Why this can happen is not clear to me, but I've seen errors in the log and we do check for it also in the equivalent functoin of intellihide.js.

Reply:
Thanks, I saw these errors too and tried too hard to find the cause. It's perhaps better to simply check for this, as you say. Addresses here: [64d7a5d6ceb0bf931b303baf5c3282e92f78072c].